website/public_adventures.py

```diff
@@ -31,67 +31,168 @@ def __init__(self, db: Database, achievements: Achievements):
-                "creator_public_profile": public_profile,
-                "date": utils.localized_date_format(adventure.get("date")),
-                "level": adventure.get("level"),
-                "levels": adventure.get("levels"),
-                "language": adv_lang,
-                "cloned_times": adventure.get("cloned_times"),
-                "tags": adv_tags,
-                "text": content,
-                "is_teacher_adventure": True,
-            }
-
-            # save adventures for later usage.
-            for _level in adventure.get("levels", [adventure.get("level")]):
-                _level = int(_level)
-                if self.adventures.get(_level):
-                    self.adventures[_level].append(current_adventure)
-                else:
-                    self.adventures[_level] = [current_adventure]
-
-            available_levels = adventure["levels"] if adventure.get("levels") else [adventure["level"]]
-            self.customizations["available_levels"].update([int(adv_level) for adv_level in available_levels])
-
     @route("/", methods=["GET"])
     @route("/filter", methods=["POST"])
     @requires_teacher
-    def filtering(self, user, index_page=False):
+    def filtering(self, user):
+        level = request.args["level"] if request.args.get("level") else "1"
+        language = request.args.get("lang", g.lang)
+        tag = request.args.get("tag", "")
+        search = request.form.get("search", request.args.get("search", ""))
+        print('\n\n\n', level, language, tag, search, '\n\n\n')
+
+        available_languages = set()
+        available_levels = set()
+        available_tags = set()
+
+        # Get all possible filters
+        lang_level_filters = self.db.get_public_adventures_filters({"field": "lang#level", })
+        for record in lang_level_filters:
+            lang, _level = record.get("value", "#").split("#")
+            available_levels.update([_level])
+            if level == _level:
+                available_languages.update([lang])
+        customizations = {"available_levels": available_levels}
+
+        tag_filters = self.db.get_public_adventures_filters({"field": "tag"})
+        for record in tag_filters:
+            _tag = record.get("value")
+            available_tags.update([_tag])
+
+        customizations = {"available_levels": available_levels}
+
+        tags = []
+        if tag:
+            toReset = request.args.get("reset")
+            if toReset:
+                # then it's the current selected tags, so remove current's tag
+                tags = [_tag for _tag in toReset.split(",") if _tag != tag]
+            else:
+                tags = tag.split(",")
+            tags = [t for t in tags if t]
+
+        # Get indexes
+        level_adventure_ids = self.db.get_public_adventures_indexes({"field#value": f"level#{level}"})
+        lang_adventure_ids = self.db.get_public_adventures_indexes({"field#value": f"lang#{language}"})
```

Comment:
I would also do a combined index of `language + level`, so that you can filter on 2 conditions at the same time. 

The problem is that if you do it by combining 3 different indexes like this, you won't be able to paginate properly.

Reply:
How would you combine them?

Reply:
I was thinking `{ "field#value": "level_lang#5_es", "date#adventure_id": "12345#abcd" }`, so you can search for a combination of level + language in one query.

That's easier and quicker than querying 2 sets and intersecting them, plus you can paginate through a single query (and you can't paginate very well through multiple queries)